Test all cities with an exact result

gonzaponte opened the pull request:
This PR adds a test to each city comparing the output for a particular configuration with the true one, which is also stored. The true output is defined as the current one. This PR is devoted to make the transition to the liquid cities (#435) as safe as possible.

gonzaponte committed:
Cache the output of database functions

gonzaponte committed:
Add test to ensure the database is being cached

Aretno committed:
Add a general histogram writer

Aretno committed:
Add and test Histogram and HistoManager classes.

Aretno committed:
Add Histogram related functions

Added plot, save histogram an join/sum functions.

Aretno committed:
Add and test monitoring functions for pmaps and rwf

Aretno committed:
Problem in test with - char

Aretno committed:
Cosmetics

Commit:
Produce and write true voxels

Aretno committed:
Modifications according to the PR review

Aretno committed:
Move get_histograms_from_file to hist_io

Aretno committed:
Added the number of sensors in wvf plots

Aretno committed:
Temporal script to run monitoring

Aretno committed:
Adapt code to reviewer petitions.

Aretno committed:
Modified style of 1D histograms (to match paper style)

Aretno committed:
Variable name change to common criteria.

Commit:
Adds tests for detsim and voxel io

Commit:
Updates code to recent version of master

Commit:
Removes unnecessary test fixture

Commit:
Adds h5 file of toy voxels to test data directory

paolafer committed:
Adapt code to change in types

Commit:
First version of channel parameter output

channel_param_io and associated test added.
Thought for generic use with any sensor type
or other numbered channel.

Updated names to be a bit more consistent with DB

improved names

added explicit sensor info in table title

Commit:
Addition of simulation of low frequency noise

Simulation of low frequency noise effects
seen in the PMTs to the sensor simulation.

Involves changes to the database to allow
parameters from the detector to be updated
whern needed.

load_db.py:
New funciton added 'PMTLowFrequencyNoise'
which reads the mapping of PMTs to
front end boxes and the frequencies
and magnitudes to be used in the simulation.

low_frequency_noise.py:
New function that randomises contributions
and returns a function to access these
according to front end board mapping

gonzaponte committed:
Add hdf5 table comparator

gonzaponte committed:
Add test to check exact result in diomira

gonzaponte committed:
Add test to check exact result in irene

gonzaponte committed:
Add test to check exact result in isidora

gonzaponte committed:
Add test to check exact result in dorothea

gonzaponte committed:
Add test to check exact result in penthesilea

gonzaponte committed:
Fix isidora

The event number was not correctly given to the mc writer

paolafer committed:
Unify type of event number across IC

paolafer committed:
Trick to reuse the same test files with the type change

Aretno committed:
Moved plotting functions to icaro folder

Aretno committed:
Tests for hist_io

gonzaponte committed:
Add test to check exact result in pmtgain

gonzaponte committed:
Add test to check exact result in sipmgain

gonzaponte committed:
Add test to check exact result in sipmpdf

jmbenlloch committed:
462  Change types in MC true info

https://github.com/nextic/IC/pull/462

[author: paolafer]

This PR is the adaption of the IC code to a problem of overflow in
nexus MC true tables.
The types of the columns of the extents table have been changed from
unsigned int or int to uint64 or int64 in order to be able to
represent a larger number of events. Strangely (to me), it turns out
that, when adding a uint64 to an int, python gives as a result a
float64 (see, for instance, this thread). Therefore, in this PR I
simply cast my uint64 variable to int, which, if I'm not wrong, is the
same as a long in python, so it should be ok.

[reviewer: jmbenlloch]

I think this update is ok. It makes uniform the type of event numbers
across the different tables and updates the table MC/extents to avoid
overflows. We need this update to be able to continue with the
development of related projects like Nexus. I have rebased this branch
to the latest master. If all tests pass I will approve and merge.

Aretno committed:
Added individual PMT plots

Aretno committed:
Added fill_pmt_var test

Aretno committed:
Review comments

gonzaponte committed:
Set method to reset `mc_info_writer`

This is needed every time the input file changes

gonzaponte committed:
Reset `mc_info_writer` automatically

gonzaponte committed:
Rename tests

This way, it is more obvious what function they test

gonzaponte committed:
Add test for the reset method

gonzaponte committed:
Use `output_tmpdir` instead of `config_tmpdir`

gonzaponte committed:
Remove unnecessary path concatenation

gonzaponte committed:
Remove unused variable

gonzaponte committed:
Fix indentation

gonzaponte committed:
Add test for automatic reset

gonzaponte committed:
Remove custom reset of `mc_info_writer` in the cities

Commit:
Changed dictionary definition to be within writer

Changed where the dictionary for a writer is
defined so that it's in the writer definition
itself. In this way errors associated with
two instances in the same job are avoided.

paolafer committed:
Add tests to show incorrect behaviour

paolafer committed:
Add check for the first file in loop

paolafer committed:
Add new test files

paolafer committed:
Keep the only variable to be reset in test

jacg commented on invisible_cities/core/testing_utils.py:
```diff
@@ -134,3 +134,19 @@ def assert_PMap_equality(pmp0, pmp1):
+        assert type(got) == type(expected)
+        try:
+            assert got == expected
+        except ValueError:
```
I struggle to see how `assert a == b` could ever raise `ValueError`

jacg committed:
Remove :unbind from use-package

It doesn't seem to be supported (any more?), and now gives errors on
startup.

jacg committed:
Replace ido-switch-buffer with helm-mini on C-x b

Commit:
New tests for individual functions

Added function to define parameter dictionary
for table definition.

Added tests for some of the individual functions
used in definition and storing of table.

gonzaponte committed:
Add tests to demontrate bugs

The sensors' waveforms are not properly matched to their IDs because
the IDs are sorted (independently from the waveforms) when read. This
is illustrated in these tests.

gonzaponte committed:
Fix bug in `build_*_responses`

Besides removing duplicates, `np.unique` sorts the input values, which
creates a mismatch between the sensor's id and its
waveform. `pd.unique` on the other hand preserves order with one
caveat: it takes into consideration only *consecutive*
duplicates (i.e. repeated values with a different one in between are
counted multiple times). Since our waveforms are always written on a
sensor-by-sensor basis, the same ID cannot be read twice in the same
peak and this tool is still useful. Two alternative
implementations (one with `np.unique` and one with
`itertools.groupby`) have also been tested and the `pd.unique` is the
fastest one, particularly for long arrays.

gonzaponte committed:
Add fix using `itertools.groupby`

gonzaponte committed:
Archive itertools implementation

gonzaponte committed:
Add fix using `numpy.unique`

gonzaponte committed:
Archive numpy implementation

Commit:
468 Fix pmap io

https://github.com/nextic/IC/pull/468

[author: gonzaponte]

The SiPMs waveforms were not properly matched to each sensor. This was
caused by `np.unique` which, besides removing duplicates, sorts the
elements. This can also be a problem in the PMTs waveforms, although
the PMTs are, in principle, always sorted.

A couple of tests have been added to demonstrate the situation and to
ensure this doesn't happen again in the future.

The solution consists of replacing `np.unique` by `pd.unique`, which
preserves the original order. There is one caveat, though: `pd.unique`
removes only consecutive duplicate elements. This is ok as our data is
stored on a sensor-by-sensor basis and there should not be
non-consecutive duplicate elements.

Two other alternative implementations have been also considered, but
the one using `pd.unique` is the fastest.

[reviewer: jmbenlloch]

This is a VERY important bug fix. The solution chosen by @gonzaponte
is the fastest one and also includes two simple tests to avoid this
situation in the future.

Commit:
Updated function description

Commit:
Generalised test to check multiple sensor entries

jmbenlloch committed:
463 Add method to reset mc writer

https://github.com/nextic/IC/pull/463

[author: gonzaponte]

The MC writer keeps track of what it has read and written from the mc
table in the input file. However, when multiple files are read in the
same city execution, the class attributes must be reset in order to
read them correctly. We are currently reseting them by hand in the
cities, but this is far from optimal.

In this PR we include a new method to the MC writer class so it
manages the attributes when a new file is given.

[reviewer: jmbenlloch]

This solves a bug when reading multiple files including several tests
to ensure everything is working ok. We also need this to start a new
montecarlo production so I will approve and merge.

Aretno committed:
Plot 3d histogramas as 2d profiles

Aretno committed:
Added monitoring plots for kdst

paolafer committed:
New test for bad behaviour when reading multiple files in one case

paolafer committed:
Fix bad behaviour when reading multiple files

jmbenlloch committed:
470 Write MC true info of multiple input files

https://github.com/nextic/IC/pull/470

[author: paolafer]

This PR fixes a bug in the MC true information writing, which only
appears in a very specific case, i.e., when a job reads multiple files
and the first event of the first file is filtered out by some
selection that is done in the job.
A test has been added to show the incorrect behaviour and the code has
been subsequently fixed.

[reviewer: jmbenlloch]

Simple and concise PR, first includes a test that shows the failure
and then a commit fixing it. Good job!

Commit:
471 Fix tornado version to 4.5.1 py36_0

https://github.com/nextic/IC/pull/471

[author: bpalmeiro]

I had some issues using Jupyter Notebook and finally, Gonzalo and I
solved by fixing tornado package version to 4.5.1 py36_0

[reviewer: gonzaponte]

Clear fix.

andLaing committed:
Changes in calibration cities for other setups

Reintroduced the copy of sensor information
from the raw files to the histogram files.
In this way a setup without reliable database
sensor mapping.

Protections introduced so that framework works
with data where there are no data from one
type of sensor.

centralised copy_sensor_table

Update to make the code work if there are no PMTs or Sipms

Fixed accidentally deleted 1 sensor type protections

Temporary patch to read files without sipmrwf

Improved sipmpdf adc_only mode

Aretno committed:
Added options in order to plot reference histograms

Added extra arguments for reference histograms (only 1d), histograms
are now normalized (optional) for the sake of comparison, stats
ploting is now an option.

Aretno committed:
Added kdst related config and test histofile

Aretno committed:
Correct RMS

gonzaponte committed:
434 Implement monitoring-related functions

https://github.com/nextic/IC/pull/434

[author: Aretno]

This PR includes a bunch of functions to manage and store histograms
for monitoring purposes. Another set of functions to pick specific
values from data structures and histogram them has also been
introduced.

[author: gonzaponte]

Lots of ugly (although necessary) functions to implement data
monitoring have been added. A review and refactor of them will
probably be necessary at some point, but it is good to go.

andLaing committed:
Added readers for writer output

andLaing committed:
Added readers to tests

andLaing committed:
Made test parameter names more PEP8 ready

andLaing committed:
Fixed another parameter style

jjgomezcadenas committed:
472 Fix emacs init compilation error

[author: jacg]

The current `init.el` fails to compile out of the box, because it uses
a feature which seems to be supported no more. The first commit in
this PR fixes that problem.

The second commit in this PR is not directly related, but as it's such
a minor thing, I'm proposing to sneak it in, in order to save opening
a separate PR for trivialities. It enables helm for buffer switching.

[reviewer: jjgomezcadenas]

Good magic indeed. Agreed to sneak the small commit in this PR.

Aretno committed:
473 Add number of entries to the stats printed in the histograms

https://github.com/nextic/IC/pull/473

[author: Aretno]

Small change to add the number of entries to the printed statistic
values when plotting histograms.

[reviewer: gonzaponte]

Clear and simple.

gonzaponte committed:
Allow events with multiple S1s in Dorothea

This adds flexibility when trying to understand new data with minor
complications.

gonzaponte committed:
Remove duplicities

gonzaponte committed:
Update fixture and test file

gonzaponte committed:
Update test

gonzaponte committed:
475 Allow multiple s1 events to be stored in Dorothea

https://github.com/nextic/IC/pull/475

[author: gonzaponte]

This PR adds the possibility of include multiple-S1 events in the kDST
produced by Dorothea.  Useful to study new data.

[reviewer: jjgomezcadenas]

This PR implements a functionality that was found necessary in the
data analysis in the previous run (the Irotea saga). Clean and simple,
with a cosistent test.

andLaing committed:
Protect changes

gonzaponte committed:
Add function to mask waveforms according to database

gonzaponte committed:
Add interface in cities

gonzaponte committed:
Mask waveforms in irene

gonzaponte committed:
Fix test

paolafer committed:
Fix bug reading the response of the last sensor written in one event

paolafer committed:
Divide one big test into two smaller ones

jmbenlloch committed:
477 Fix bug reading the MC response of the last sensor written

https://github.com/nextic/IC/pull/477

[author: paolafer]

The function that reads the Monte Carlo waveforms of the sensors was
not reading the last sensor in each event. I have added a test which
reveals this bug and fixed the code accordingly.

[reviewer: jmbenlloch]

This PR fixes a bug when reading MC info and includes a test that
shows everything is working ok.

andLaing committed:
Added tests for changes

sipmpdf_test parametrized to test both
possible cases: adcPDFs and pePDFs

Added test for copy sensors

gonzaponte committed:
476 Nullify waveforms of masked sensors

https://github.com/nextic/IC/pull/476

[author: gonzaponte]

In this PR the waveforms of the masked sensors are nullified (set to
0). This way, they can never be used for anything.

I think we've never fully masked the dead sensors. My feeling is that
we just masked them in some particular scenarios (corona algorithm),
but not generally (when corona is configured to perform a barycenter,
for instance).

[reviewer: jjgomezcadenas]

This PR kills a monster that has been killed, I think a number of
times and then has resurrected. Hopefully it will stay dead from now
on.

gonzaponte committed:
Combine S1 and S2 times correctly to compute the correct z and dt

gonzaponte committed:
Save nan for z and dt when no S1 is found in the event

gonzaponte committed:
Remove empty s1/s2 filter

gonzaponte committed:
Fill default values in case of an empty S1 or S2

gonzaponte committed:
Update test

gonzaponte committed:
479 Allow the saving of events with 0 S2 or 0 S2 in dorothea

https://github.com/nextic/IC/pull/479

[author: gonzaponte]

Allow the saving of events with 0 S2 or 0 S2 in dorothea.
Useful to keep track of detection efficiencies.

[reviewer: jjgomezcadenas]

All straight forward and badly needed for immediate analysis of data
popping out for the machine.

jjgomezcadenas committed:
480 Flatten mask in display_matrix

https://github.com/nextic/IC/pull/480

[author: jjgomezcadenas]

The display function display_matrix takes an optional mask array that
allows to plot only those values of the matrix accepted by the
mask. This feature had not been previously tested, and thus we had not
noted that the mask had to be flattened. This is now fixed.

[reviewer: gonzaponte]

Straightforward fix. Approved.

jjgomezcadenas committed:
Add property interval to minmax

A minmax can be used to specify a range. Property interval returns
such a range as a pair (min, max)

jjgomezcadenas committed:
Test minmax property interval

andLaing committed:
Added no deconvolution option to pmtgain

andLaing committed:
fixed error in new mode

Aretno committed:
482 Reduction of memory usage for monitor plotting

https://github.com/nextic/IC/pull/482

[author: Aretno]

Quick fix to reduce memory usage for histogram plotting for the
monitoring

[reviewer: gonzaponte]

Nice. Desperately needed to avoid crashes on a daily basis.

Aretno committed:
483 Fix bug introduced in #482

https://github.com/nextic/IC/pull/483

[author: Aretno]

If-related fuck-up, fixed.

[reviewer: gonzaponte]

Ok.

andLaing committed:
fixed tests

paolafer committed:
Add test to show bug in assigning bin wodth to sensors.

paolafer committed:
Retrieve correct sensor ID for waveform to be saved to dictionary

paolafer committed:
Read also units for sensor binning parameters

paolafer committed:
Remove comments

paolafer committed:
Add picosecond to time units

jmbenlloch committed:
486 Fix bug last written sensor

https://github.com/nextic/IC/pull/486

[author: paolafer]

There was a bug in the code of read_mcsns_response: when storing a new
Waveform object, the bin width of the following one was taken. A test
has been added to show this bug, and the code has been fixed. In this
process, I've realized that the bin width was actually being saved
without units, so I have also added this feature. The test added can
serve also as a test for the units being saved correctly.

[reviewer: carmenromo, jmbenlloch]

We have reviewed the code, there was indeed a bug and it has been
successfully fixed including a test that shows it.

Aretno committed:
491 Silence flood of NaturalNameWarring

[Author: Arento]

Close #484

[reviewer: jacg]

The `NaturalNameWarnings` have been silenced. There is still an
excessive number of hypothesis-related warnings, but those are caused
by an unrelated issue, and need a completely unrelated fix.

I hereby approve this PR.

jmbenlloch committed:
Remove whitespace

jmbenlloch committed:
474 Changes in calibration cities for other setups

https://github.com/nextic/IC/pull/474

[author: andLaing]

The changes to protect against data with different configurations,
particularly with only one type of sensor were necessary for the work
that is being done at IFIC at the moment but are probably insufficient
for the future when we will have to run over several different
detector configurations.

Reintroduced the copy of sensor information
from the raw files to the histogram files.
In this way a setup without reliable database
sensor mapping.

Protections introduced so that framework works
with data where there are no data from one
type of sensor.

centralised copy_sensor_table

Update to make the code work if there are no PMTs or Sipms

Fixed accidentally deleted 1 sensor type protections

Temporary patch to read files without sipmrwf

Improved sipmpdf adc_only mode

[reviewer: gonzaponte]

We need to seriously think how to manage different detector
configurations with our code automatically, but this patch is
perfectly fine for the time being.

andLaing committed:
Fixes so that both ZS data and MC can be processed

Designed to avoid bad baseline calculations
when in ZS mode and to bring MC diomira ouput
in line with ZS data format.

jjgomezcadenas committed:
Throw an exception if the table is not correctly read

jjgomezcadenas committed:
Add an excecption to describe a DST table not correctly read.

jjgomezcadenas committed:
Add a "wrong file" in the return of KrMC_kdst fixture.

This is needed to test the case in which one accidentally reads a
wrong file. The case occurs, for example, if one is reading dst files
with glob, and a file of different type than expected is present in
the directory.

jjgomezcadenas committed:
Adapt test, KrMC_kdst is now a tuple.

jjgomezcadenas committed:
Add a file of "wrong type" to test reading DSTs

andLaing committed:
removed print and implemented wrapper

jjgomezcadenas committed:
Add assert to test.

jmbenlloch committed:
Remove whitespace

jmbenlloch committed:
Mark test as flaky

jmbenlloch committed:
489 Fixes so that both ZS data and MC can be processed

https://github.com/nextic/IC/pull/489

[author: andLaing]

Designed to avoid bad baseline calculations
when in ZS mode and to bring MC diomira ouput
in line with ZS data format.

[reviewer: jmbenlloch]

This is a required update to make MC and data alike when using zero
suppression. Needed update to test the new configuration at Canfranc.

jmbenlloch committed:
481 A property interval to minmax

https://github.com/nextic/IC/pull/481

[author: jjgomezcadenas]

A minmax can be used to specify a range. Property interval allows to
return the minmax as a pair for that specific purpose.

[reviewer: jmbenlloch]

Simple change including a test. Approved!

jjgomezcadenas committed:
Remove commented code

The proposed behavior of the function when an exception of type
NoSuchNodeError is found is simply to print an error message, to
inform the user of the existence of an offending file and then ignore
the file. This is a case where it is justified not to abort the concat
of good files (which can take many minutes) because a spurious or
corrupted file is found. No need to throw a new exception either,
although it could be done if one wanted to let the user decide what to do.

jacg committed:
Update pinned pcakage versions to most recent

jacg committed:
Add Python 3.7 to Travis matrix

jacg committed:
source activate -> conda activate

jacg committed:
Allow Python 3.7 builds to fail on Travis

We are not relying on Python 3.7 yet. It's there just so that we can
see when all our required packages become available.

jacg committed:
Bump conda version

jacg committed:
Add 3.7 to bash completion options

jacg committed:
Adapt paolina to evolution of NetworkX interface from v1 to v2

`all_pairs_dijkstra_path_length` has become lazy: it used to return a
dictionary. Now it returns an iterator.

Graph query methods have become lazy too, so `nodes_iter` becomes the
default `iter`

jacg committed:
Rephrase find_extrema_and_length

jacg committed:
Heed FutureWarnings about DataFrame.groupby()

The meaning of tuples passed to groupby is going to change in the
future: they will no longer specify multiple keys, but will be
understood to be a single key themselves. To specify multiple keys, a
list should be used instead.

Aretno committed:
Reduce number of non-valid histo examples for tests

Every now and then a HealthCheck error in this direction appeared (only a
few valid examples out of 50), this should address this issue at the
cost of slower test generation.

jjgomezcadenas committed:
458 Truevoxels

https://github.com/nextic/IC/pull/458

[author: jerenner]

Adds code to generate true voxels from Nexus MC hits:

+ city detsim takes Nexus input, applies energy smearing and diffusion
  to the hits, and outputs true voxels in HDF5 format (see VoxelsTable
  in `evm/nh5.py`);

+ voxel I/O is implemented in `io/voxels_io.py`, including a writer for
  true voxels and a function to load true voxels

[reviewer: jjgomezcadenas]

Yes, this is true [there is a test, and the underlying functionality
is already dested elsewhere - ed].  On this grounds I believe we can
approve this PR.

Aretno committed:
Fix error in a test

jacg committed:
Turn work_in_python_version into a one-stop solution

Setting up the environment before working on or using IC was a bit of
a PITA because you were required to use different `manage.sh`
commands, depending on whether

+ conda is already installed
+ the cond environment you want to use already exists

The `manage.sh` function `work_in_python_version` has been mae more
intelligent: it now discovers which of these steps need to be taken,
and should be usable as a single entry point to getting going with IC.

This commit also contains the beginning of dependency versioning:
the IC conda environments will now have names which include a version
stamp (rather than just the Python version being used). This should
allow us to switch to an environment which contains the dependency
versions which were in place at the time any given version of your
code was committed.

jacg committed:
Run tests in parallel by default work_in_python_version

jacg committed:
Add work_in_python_version_no_test

By default we want the tests to run, to highlight pre-existing
problems when someone starts to work in/on IC. However, as this work
is leading up to automatic environment switching when historical
commits are checked out, we do not want to run all the tests on each
checkout. So we need a version that switches environment without
running the tests.

jacg committed:
Ensure that PYTHON_VERSION is set sooner in manage.sh

jacg committed:
Make manage.sh testing automatically choose appropriate env

jacg committed:
Prevent run_tests_par from exiting shell

jacg committed:
Deduce PYTHON_VERSION from env if not passed as arg

The manage.sh testing functions do not accept a python version as an
argument. But they need to know this version, now that they need to
switch automatically to the environment which corresponds to the
checked out code.

jacg committed:
Update quickstart guide

Aretno committed:
Added generic dictionary key filtering

Aretno committed:
Filter DSTs by event id.

Aretno committed:
Filter pmaps by event id.

jjgomezcadenas committed:
Replace print by warning.

jmbenlloch committed:
446 Cache database

https://github.com/nextic/IC/pull/446

[author: gonzaponte]

The functions of the database are decorated with functools.lru_cache,
which keeps the result of such functions calls associated with their
parameters so they are not called every time. This assumes the output
of the function is unchanged when called with the same parameters, as
is the case for the database. This comes in response to #445 and to
 #435, which needed/considered caching the database to improve
performance.

[reviewer: jmbenlloch]

I have rebased this old branch and checked the code. A simple change
that can improve DB times by a great factor (test shows a factor
1e6).

jmbenlloch committed:
497 Add functions to filter based on Event ID

https://github.com/nextic/IC/pull/497

[author: aretno]

This PR adds two functions to filter data based on the event id, one
for DSTs and another for pmaps. The latest one uses a more generic
function (also added in the PR) which filters dictionaries based on
their keys. There was a previous dictionary filter but only on the
values, this function has been renamed to clearly state the difference
between both cases.

[reviewer: jjgomezcadenas]

Excellent! the PR is concise and the tests are satisfactory, the
functionality is really necessary.

carmenromo committed:
Fix deconv_pmt docstring

carmenromo committed:
Fix type hint in mcinfo_io

carmenromo committed:
Remove liar comments in TlMC_hits and TlMC_hits_skipping_NN

jmbenlloch committed:
Cosmetics

jmbenlloch committed:
464 First version of channel parameter output

https://github.com/nextic/IC/pull/464

[author: andLaing]

channel_param_io and associated test added.
Thought for generic use with any sensor type
or other numbered channel.

Updated names to be a bit more consistent with DB

improved names

added explicit sensor info in table title

[reviewer: jmbenlloch]

This PR includes a new functionality with the corresponding tests. I
have added some minor cosmetic corrections. I think it is ready for
approval.

jmbenlloch committed:
490 Dst IO

https://github.com/nextic/IC/pull/490

[author: jjgomezcadenas]

This PR introduces a protection in the function that reads dorothea
DSTs, throwing an exception if the operation is not successful. This
protection is necessary to prevent the case in which a spurious file
appears in a directory containing DSTs (and being read out using
glob). Without the exception, the function will read files until it
finds the wrong file, then die ungracefully, and the concatenation of
files will not occur.

[reviewer: jmbenlloch]

Simple update including a test.

jjgomezcadenas committed:
Add further protections to load_dst.

Function load_dst() can fail also due to corrupted files or non
existing files. In all the cases the action is to intercept the
excpetion and send a warning.

jjgomezcadenas committed:
Modify test to account for new warnings.

jjgomezcadenas committed:
Fixed a bug in conftest.

Name of wrong_file was not being properly passed and thus the expected
warning was not generated.

jjgomezcadenas committed:
Improve warnings.

Warnings defined as UserWaning and messages made clearer.

jjgomezcadenas committed:
Improve tests.

Warnings now catched properly, two different tests for the two common warnings.

Commit:
initial import to pass through MC truth generator info in HDF5 file. work in progress

Commit:
add three data files to test copying of MC generator info

jjgomezcadenas committed:
Add comment to function.

jmbenlloch committed:
503 Kdst IO

https://github.com/nextic/IC/pull/503

[author: jjgomezcadenas]

This PR improves over a previous PR in which basic protections were
added to load_dst function, by adding two additional cases in which
exceptions are intercepted and warning send without interrupting
execution.

[reviewer: jmbenlloch]

Improved version on #490 including new tests.

Commit:
add test_copy_mc_generator_info test

jmbenlloch committed:
504 Copy trigger information from RD to pmaps

https://github.com/nextic/IC/pull/504

[author: jmbenlloch]

This PR copies the trigger information written by the decoder in the
raw waveforms files. The structure is the following:

    /Trigger/events is an EArray with the information about the
    channel(s) that triggered the event. The first dimension is the
    event and the second one the PMTs, which are sorted in the same
    order as the waveforms, DataPMT from load_db, etc.

    /Trigger/trigger is a table with the trigger type given by the
    DAQ. Under normal operation there could be two codes:
    1 (corresponding to trigger 1) and 9 (corresponding to trigger
    2). There is one entry per event in this table.

With the changes in this PR, Irene will propagate the information to
the pmaps files. Two new tests are added to check both of the previous
data structures are properly propagated.

It is important to say that this functionality has been implemented in
a backward-compatible way. Irene will still be able to process files
that has no trigger information.

[reviewer: jjgomezcadenas]

This PR extends the functionality of IC to incorporate trigger info to
the reco chain. All the necessary methods are provided. most of the
time extending existing structures. Two tests are provided to ensure
consistency.

jmbenlloch committed:
505 Split output files in irene in case of double trigger

https://github.com/nextic/IC/pull/505

[author: jmbenlloch]

This PR adds to Irene the ability of writing two separate output files
in case of double trigger operation. Since the changes included here
are only relevant for Irene, most of the new code has been added to
irene.py instead of base_cities.py to avoid making changes that can
affect other unrelated cities.

Four new parameters are to Irene:

   -split_triggers: Boolean, tells Irene whether to split the output
    or not in two files. This parameter is mandatory.
   -trg1_code: Int, trigger code for trigger1 given by the DAQ. Now
    should be 1, no changes foreseen in the near future.
   -trg2_code: Int, trigger code for trigger2 given by the DAQ. Now
    should be 9, no changes foreseen in the near future.
   -file_out2: Path, output filename for events of trigger 2.

The three latter parameters are only needed if split_triggers is True.

With this PR Irene will be able to write all the events to one file or
classify the events based on their trigger type and write each of them
to the proper file. The usual file_out will be for trigger 1 and
file_out2 for trigger 2.

If for some reason there is an event with a trigger type different
from trg_code1 or trg_code2 a warning will be printed.

A new test has been added to check all that is done properly.

[reviewer: jjgomezcadenas]

Continuation of previous PR, allows splitting the stream in two
independent streams, according to trigger.

Commit:
deal with case where hdf5 file does not have MC generators information, and revert back one change in read_mcinfo_evt

Commit:
remove comments

Commit:
removed unused variable

paolafer committed:
506 Copy generator info in HDF5 MC files

[author: msorel]

This is to pass through MC generator info through the various IC
processing steps. This is the 'generators' dataset in the 'MC' group,
and contains 'evt_number', 'atomic_number', 'mass_number' and 'region'
information, that changes event-by-event.

This additional information is necessary for the background MC
production, mixing many background generators in the same file.

The test 'test_copy_mc_generator_info' has been created. It is meant
to cover three cases:

1. mcfile_withgeneratorinfo: MCRD file where 'MC' group has
   'generators' dataset with non-zero dimension, equal to number of
   events in file. Produced with GATE version v1_03_00 or later, and
   from nexus file where GATE event string store contains
   "/Generator/IonGun/atomic_number", "/Generator/IonGun/mass_number"
   and "/Generator/IonGun/region" information

2. mcfile_withemptygeneratorinfo: MCRD file where 'MC' group has
   'generators' dataset with zero dimension. Produced with GATE
   version v1_03_00 or later, and from nexus file where GATE event
   string store does NOT contain "/Generator/IonGun/atomic_number",
   "/Generator/IonGun/mass_number" and "/Generator/IonGun/region"
   information

3. mcfile_withoutgeneratorinfo: MCRD file where 'MC' group has NO
   'generators' dataset. Produced with GATE version prior to v1_03_00

[review: paolafer]

Nice workaround, Michel! It looks like it works fine.

jmbenlloch committed:
494 Reduce number of non-valid histo examples for tests

https://github.com/nextic/IC/pull/494

[author: aretno]

Every now and then a HealthCheck error in this direction appeared (only a
few valid examples out of 50), this should address this issue.

[reviewer: jjgomezcadenas]

In my opinion this PR is mature enough, after the long discussion to
be approved.

paolafer committed:
Add check for zero particles or zero hit files.

paolafer committed:
Add tests and files.

paolafer committed:
Add error message if accessing sensor response when not available

carmenromo committed:
Change List to Sequence type in mcinfo_io.py

paolafer committed:
Add files for tests.

Commit:
Adds new fields for corrected charge and energy to Hit and Cluster

jjgomezcadenas committed:
Fix signal.decimate

The change from Scipy 0.2 to Scipy 1.1 caused a nasty error due to a
parameter (n) which changed its default value. This is fixed by
setting explicitely n to 30 (which was the default value in 0.2).

jjgomezcadenas committed:
Add two tests to document the decimate error.

One test shows the problem and the other one its solution.

paolafer committed:
Add specific exception for accessing particle info where there's not

Commit:
Adds hit property Zc

Commit:
Updates tests to accomodate Hit property Zc

paolafer committed:
Make the error message more general.

paolafer committed:
Economize the code and add a comment to make test clear.

jjgomezcadenas committed:
Remove test_show_scipy_nasty_feature.

Too verbose and difficult to understand. Instead,
test_show_signal_decimate_signature should be enough to document the issue.

jjgomezcadenas committed:
replace which conda by conda --version

Command which conda does not work in MacOS. Using conda --version
works and returns the current version of conda, which may be
nice. Without this, the script attempts to install conda every time in MacOS.

jacg committed:
Stop failing to run parallel tests in work_in_python_version

The `manage.sh` command line argument sometimes specifies the python
version, sometimes the number of processors to use for parallel
testing. `work_in_python_version` accepts a python version, but
internally it calls run_tests_par, which requires the command line
argument to specify the number of processors. This conflict resulted
in the parallel tests not being run at all.

This is a quick hack to fix the problem as quickly as possible.

jjgomezcadenas committed:
496 Dependency versioning

[author: jacg]

An attempt at addressing #495.

The idea is:

+ Include a dependency configuration tag in a committed
  file (manage.sh for now).

+ Povide manage.sh commands which

  - create conda environments which have the tag in their name and
    contain the correct set of dependencies

  - switch conda environment

+ Make the manage.sh testing commands automatically switch to the
  conda env which is appropriate for the currently checked out code.

+ The environments are made lazily: no environment is created until it
  is needed, which will only happen if

  - A commit which needs that environment is checked out, and
  - The user asks for the environment to be created, ether
    + explicitly, or
    + implicitly, by using manage.sh to run tests.

The tags are, for now, of the form YYYY-MM-DD, but we could use relase
tag names instead. It might be a good idea to update the dependencies
on every release: that way we'll never get too far behind the march of
progress.

An example: If you clone and install the repo containing this branch
today, you will create a conda env called IC-3.6-2018-06-30. If you
then check out a commit which corresponds to an earlier dependency
configuration, as soon as you run the tests with source manage.sh
run_tests(_par), a second conda environment called IC-3.6-2018-05-03
will be created and activate implicitly. Checking out the tip of the
branch once more, and running the tests (with manage.sh) will
automatically switch to the first environment.

The whole lot needs to be cleaned up, made more robust, and the
database should be updated automatically too. We also need to think
about how to deal with commits which predate the addition of this
feature.

I can imagine that better approaches could be invented. Let me know
what you think.

edit:

It seems that we don't have the manpower to think about this carefully
and the only way to prevent this from being ingnored and festering
forever, is to approve and commit the first thing that looks like it
works.

[reviewer: jjgomezcadenas]

I have tried the system and works. It may create a bit of hassle when
conflicting with the old version of conda, but it can be easily
solved.

jacg committed:
Bump versions not available in Python 3.7

The old versions of hypothesis, pandas and pymysql are not available
in the standard conda channel for Python 3.7, so these need to be
bumped for Python 3.7 to work at all.

jacg committed:
Bump versions which don't break anything

Bumping Numpy from 1.14.5 to 1.15.0 breaks `test_irene_split_trigger`,
so we'll keep that one back, and deal with it in a separate PR.

Cython, notebook, pytest and sphinx are available in their old
versions for Python 3.7, but me might as well take this opportunity to
keep up with the latest and greatest.

jacg committed:
Bump Numpy version, breaking test_irene_split_trigger

We were using Numpy 1.14.5. 1.15.0 is available, but
`test_irene_split_trigger` breaks if we use the new version of
Numpy. This commit demonstrates that fact.

jacg committed:
Tolerate warnings which don't concern us

`test_irene_split_trigger` used to check that exactly one
warning (with a very specific message) arose is a given
situation. When other, orthogonal, warnings appeared, the test failed.

The solution is to ensure that there is at least (rather than exactly)
one warning, and to check that the required message appears exactly
once in the collection of warnings that do appear.

jacg committed:
515 Create new environment if necessary, when switching

[author: jacg]

This addresses a bug in #496: it (sometimes?) crashed when needing to
switch to an environment that had not been created yet.

The fix consits of ensuring that, if the environment does not yet
exist, it is created before the switch is attempted.

(I have no idea how to write automated tests for the stuff in
manage.sh, especially the ones related to switching environment
... which makes working on this very painful and error-prone.)

[reviewer: jjgomezcadenas]

The code is straight forward and can be approved right away. I share
the worry about automated tests here, but let's move this onwards.

jjgomezcadenas committed:
510 Adds new fields for corrected charge and energy to Hit and Cluster

[author: jerenner]

https://github.com/nextic/IC/pull/510

This pull request creates four new fields, two in class `Hit` and two
in class `Cluster`:

+`Ql`: lifetime-corrected charge
+`El`: lifetime-corrected energy
+`Qc`: fully-corrected charge (lifetime + x,y)
+`Ec`: fully-corrected energy (lifetime + x,y)

The field `Ql` and `Qc` are added to the `Cluster` class, and the
fields `El` and `Ec` are added to the `Hit` class. These fields are
automatically initialized to -1 if not specified and are intended to
be filled later when corrections are made.

[reviewer: jjgomezcadenas]

The new fields are needed and tests show functionality.

andLaing committed:
Add padding around above threshold SiPM samples

Makes the baseline calculation more reliable
and MC more similar to data.

Fixes bug noticed in MC runs.

jjgomezcadenas committed:
493 Keep an eye on Python 3.7

[author: jacg]

Added Python 3.7 as an allowed failure to the Travis build matrix. Like this we will know when all our dependencies are available on conda in Python 3.7.

I have also taken this opportunity to:

+ Update the pinned versions of our dependencies to the most recent ones

+ Update conda to the most recent version, and change the acivation
  command to the one which has been available and recommended since
  4.4 (#455).

[reviewer: jjgomezcadenas]

python 3.7 passes all tests in my machine. As an extra check I have
run a complex analysis notebook that performs all kind of fits to the
lifetime of the detector and the results are identical in 3.6 and
3.7. Thus, it seems that the upgrade is a success.

I recommend keeping both versions of python for a short while. The
current software in production is very stable and I am reluctant to
introduce any changes which are not needed in the prod version, but
the dev version of the software can already go to the bumped-package,
3.7 python. Over the next couple of weeks we can launch a test
production of kdsts with dev and verify that everything is fine.

However I would recommend all developers to move to the bumped 3.7
version, so that once we move the dev (3.7) to prod, we don't need to
support 3.6 anymore (don't see the point, once we have checked
everything is fine and transition has been completed.

jmbenlloch committed:
514 Fix Irene bug caused by Numpy upgrade

https://github.com/nextic/IC/pull/514#issuecomment-408897527

[author: jacg]

We were using Numpy 1.14.5. 1.15.0 is available, but
`test_irene_split_trigger` breaks if we use the new version of Numpy.

(No solution provided yet: we're only demonstrating the bug, so far:
WIP)

(This sits on top of #493, so don't merge it until that one is done.)

[reviewer: jmbenlloch]

You're right. This is approved, the bug has been successfully fixed.

andLaing committed:
Slight change to clip commands

jacg committed:
Skip test_inverse_cdf_index_hypothesis_generated

This test is making half of our Travis builds fail, and we do not have
manpower available to get to the bottom of the problem right now. As
it's really interfering with PR approval and merging, we're switching
it off for now.

I am taking the liberty to push this without PR and review.

paolafer committed:
Add columns name for empty dataset.

paolafer committed:
Eliminate unused variables.

paolafer committed:
Swap if and while statements.

andLaing committed:
Rebased and tests added

Commit:
511 Add protections for true MC reading

[author: paolafer] 

Inspired by Michel's last PR, I fixed a bug, which caused a crash if
the MC true information was accessed, in files where no true hits were
saved. A few more exceptions have been added, to give meaningful error
messages in the following cases:

+In a file with no sensor response information, the user tries to
access sensor responses.
+In a file with only sensor response information (no hits, neither
particles stored), the user tries to access true info. This is the
case, for instance, of the nexus output of look-up table generation.

[reviewer: msorel]

I agree with your latest changes @paolafer. This PR improves robustness
of code and can be merged.

Commit:
fix copy of MC extents dataset for the case where filtered out events exist, and event numbers are not necessarily in increasing order

Commit:
498 Fix comments, doctrings and type annotations

[author: carmenromo]

There are various places where comments have diverged from reality,
and therefore mislead the reader. This PR fix some comments, doctrings
and type annotations and also serves to remind other users to
contribute information about what else needs to be fixed.

closes #419

[reviewer: bpalmeiro]

Great job!

andLaing committed:
Fixed error with thresholds

Addition was not performed cleanly and
thresholds increased with time.

mmkekic committed:
524 Change load_pmaps function to improve performance

https://github.com/nextic/IC/pull/524

[author: mmkekic]

`load_pmaps` is very slow for large pmaps files. I believe due to the
lookup times in dataframe. I propose minimal change that uses groupby
object so that `load_pmaps` scales linearly with the number of events.
The comparison of current and proposed load_pmaps performance for 3
different files (~300, ~10k, ~40k events) is shown in figure (see
online).

[reviewer: jjgomezcadenas]

Boosts performance, smart and concise.

Commit:
fix bug with copy of MC generator information: when an event is filtered out, the row being copied is not the last one

Commit:
expand the test to check the copy of mc info for non consecutive events. the expansion covers the case where event numbers in the file are not ordered in ascending order. This is the case for background MC files, typically

jjgomezcadenas committed:
516 Add padding around above threshold SiPM samples

https://github.com/nextic/IC/pull/516

[author: andLaing]

Makes the baseline calculation more reliable and MC more similar to
data.

Fixes bug noticed in MC runs.

This didn't leave enough information to calculate the baselines so it
tended to be overestimated, sometimes by a large amount. The change is
to take a bit before and after each identified above threshold sample
which is also more similar to what the FPGA does for the SiPMs in zero
suppression mode. As a default I've put ±50 samples in the example
configs like the current settings of the Java do for data although the
thresholds are not identical. This results in larger output files for
the same data and slower running (not by much I don't think but I've
only checked 5 events)

I ran over a full MC data file of 9200 events and the basic plots look
good.

[reviewers: paolafer,  msorel, jjgomezcadenas]

paolafer: andLaing and myself have passed a bunch of events through
diomira and irene, and the bug seems solved: the SiPM waveforms looks
fine (while before they were not). Therefore, from the point of view
of functionality it seems ok.

msorel: I want to confirm that i am happy with MC outcome of this
branch. By processing some background MCRD events through diomira,
irene, dorothea, penthesilea with this branch, I see, in short all
good. We will keep validating MC with more in-depth studies and
reportresidual oddities. But this is good enough to be merged, as far
as I am concerned.

jjgomezcadenas: Extensively tested first by paolafer then by msorel,
with consistent results.

Commit:
add test_mc_info_writer_generatoroutput_non_consecutive_events

andLaing committed:
Clarified tests

andLaing committed:
fixed typo

Commit:
change test_mc_info_writer_generatoroutput_non_consecutive_events so that it actually fails with current mcinfo_io.py code in nextic:master, while it passes for mcinfo_io.py in msorel:improve-mc-copy

paolafer committed:
500 Addition of simulation of low frequency noise

[author: andLaing]

Simulation of low frequency noise effects seen in the PMTs to the
sensor simulation.

Involves changes to the database to allow parameters from the detector
to be updated whern needed.

load_db.py:
New funciton added 'PMTLowFrequencyNoise' which reads the mapping of
PMTs to front end boxes and the frequencies and magnitudes to be used
in the simulation.

low_frequency_noise.py:
New function that randomises contributionsand returns a function to
access these according to front end board mapping.

[author: paolafer]

The code looks fine to me, and the tests added verify that the results
are consistent with what is given as an input.

paolafer committed:
525 Improve mc copy

[author: msorel]

This PR is for fixing the copying of MC truth information, in
particular the MC generator information on decaying isotope and decay
volume in the detector. In the master branch, the copy is wrong if the
two following conditions are met at the same time:

+Some event is being filtered out between input and output. This can
be the case, for example, in irene, dorothea or penthesilea cities.

+Event numbers are not in increasing order in the input file. This can
be the case, for example, in background MC files.

[reviewer: paolafer]

Ok, I think that this PR is ready to be approved!

paolafer committed:
Fix (again) type hints - reword.

jjgomezcadenas committed:
Add seaborn to our environment.

Seaborn is a high level, very useful, statistics visualization library
based on matplotlib and making extensive use of numpy and pandas. It
deserves to live in our environment.

jjgomezcadenas committed:
Changed the tag to today's date

jacg committed:
Fix specification of seaborn package in manage.sh

Package name and version should be separated by an equals sign, rather
than a minus.

jacg committed:
Apply JCK-1 to package list in manage.sh

The bug fixed in the last commit (using `-` rather than `=` in a
package specification) would almost certainly not hove been introduced
in the first place, if this had been aligned properly.

jjgomezcadenas committed:
529 Fix (again) type hints

[author: paolafer]

Apparently PR #498 didn't fix correctly the type hints of
`mc_info.py`. This PR corrects that.

[reviewer: jjgomezcadenas]

Easy enough.

paolafer committed:
Add voxel size tolerance to test

paolafer committed:
Set correct size to voxels

paolafer committed:
Fix typos in docstring

jacg committed:
530 Add seaborn to our environment

[author: jjgomezcadenas]

Seaborn is a high level, very useful, statistics visualization library
based on matplotlib and making extensive use of numpy and pandas. It
deserves to live in our environment.

[reviewer: jacg]

This PR adds a new package, which in has no impact on the rest of the
project other than slightly increasing our build time. If it is useful
for analysis, let's go ahead.

jacg committed:
Replace docstring with comment

paolafer committed:
Add test to show bug in setting voxel size

paolafer committed:
Add comment to clarify the meaning of the variable.

jacg committed:
Rephrase test

jacg committed:
Vectorize the last rephrasing

jacg committed:
532 Set correct size to voxels

[author: paolafer]

I've noticed that in the current code of voxelize_hits in
paolina_functions, the voxels are given as size the dimensions
requested by the client. However, the code actually changes slightly
these dimensions, in order to fit better the hits, event by
event. Therefore, the make_track_graphs algorithm assumes the wrong
size when building the tracks.
This PR fixes this issue.

[reviewer: jacg]

A clear mistake in the flexible mode of voxelize_hits has been
identified. A test demonstrating the problem has been added. The
problem has been fixed and the test passes.

carmenromo committed:
537 Fix redundancy in load_db file

https://github.com/nextic/IC/pull/537

[author: carmenromo]

This PR removes dbfile parameter because it was an useless variable.

[reviewer: jmbenlloch]

Simple change that will keep our code a little bit cleaner. Approved!

gonzaponte committed:
Remove annoying print that adds noise to dorothea's output
